File: source/jugglinglab/util/ParameterList.java
// ParameterList.java
//
// Copyright 2019 by Jack Boyce (jboyce@gmail.com)

package jugglinglab.util;

import java.text.MessageFormat;
import java.util.ArrayList;
import java.util.ResourceBundle;
import java.util.StringTokenizer;


public class ParameterList {
    static final ResourceBundle guistrings = jugglinglab.JugglingLab.guistrings;
    static final ResourceBundle errorstrings = jugglinglab.JugglingLab.errorstrings;

    protected int size;
    protected ArrayList<String> names;
    protected ArrayList<String> values;

    public ParameterList() {
        size = 0;
    }

    public ParameterList(String source) {
        this();
        this.readParameters(source);
    }

    public void addParameter(String name, String value) {
        if (size == 0) {
            names = new ArrayList<String>();
            values = new ArrayList<String>();
        }
        names.add(name);
        values.add(value);
        size++;
    }

    public String getParameter(String name) {
        for (int i = size - 1; i >= 0; i--)
            if (name.equalsIgnoreCase(getParameterName(i)))
                return getParameterValue(i);
        return null;
    }

    public String removeParameter(String name) {
        for (int i = size - 1; i >= 0; i--) {
            if (name.equalsIgnoreCase(getParameterName(i))) {
                size--;
                names.remove(i);
                return values.remove(i);
            }
        }
        return null;
    }

    public String getParameterName(int index) {
        return names.get(index);
    }

    public String getParameterValue(int index) {
        return values.get(index);
    }

    public int getNumberOfParameters() {
        return size;
    }

    public void readParameters(String source) {
        if (source == null)
            return;

        StringTokenizer st1 = new StringTokenizer(source, ";");

        while (st1.hasMoreTokens()) {
            String str = st1.nextToken();
            int index = str.indexOf("=");
            if (index > 0) {
                String name = str.substring(0, index).trim();
                String value = str.substring(index + 1).trim();
                if (name.length() != 0 && value.length() != 0)
                    addParameter(name, value);
            }
        }
    }

    public String toString() {
        String result = "";

        for (int i = 0; i < size; i++) {
            if (i != 0)
                result += ";";
            result += getParameterName(i) + "=" + getParameterValue(i);
        }

        return result;
    }

    // Utility function to throw an appropriate error if there are parameters
    // left over after parsing.
    public void errorIfParametersLeft() throws JuggleExceptionUser {
        int count = getNumberOfParameters();

        if (count == 0)
            return;
        else if (count == 1) {
            String template = errorstrings.getString("Error_unrecognized_param");
            Object[] arguments = { getParameterName(0) };
            throw new JuggleExceptionUser(MessageFormat.format(template, arguments));
        } else {
            String template = errorstrings.getString("Error_unrecognized_params");
            ArrayList<String> names = new ArrayList<String>();
            for (int i = 0; i < count; i++)
                names.add(getParameterName(i));
            Object[] arguments = { String.join(", ", names) };
            throw new JuggleExceptionUser(MessageFormat.format(template, arguments));
        }
    }
}

File: composeApp/src/commonMain/kotlin/jugglinglab/util/ParameterList.kt
//
// ParameterList.kt
//
// Copyright 2002-2026 Jack Boyce and the Juggling Lab contributors
//

package jugglinglab.util

import jugglinglab.composeapp.generated.resources.*

class ParameterList() {
    var numberOfParameters: Int = 0
        private set
    private lateinit var names: ArrayList<String>
    private lateinit var values: ArrayList<String>

    @Throws(JuggleExceptionUser::class)
    constructor(source: String?) : this() {
        readParameters(source)
    }

    // Return true if parameter already existed, false if it was new.
    fun addParameter(name: String, value: String): Boolean {
        if (numberOfParameters == 0) {
            names = ArrayList()
            values = ArrayList()
        }

        for (i in numberOfParameters - 1 downTo 0) {
            if (name.equals(getParameterName(i), ignoreCase = true)) {
                values[i] = value
                return true
            }
        }

        names.add(name)
        values.add(value)
        ++numberOfParameters
        return false
    }

    fun getParameter(name: String): String? {
        for (i in numberOfParameters - 1 downTo 0) {
            if (name.equals(getParameterName(i), ignoreCase = true)) {
                return getParameterValue(i)
            }
        }
        return null
    }

    fun removeParameter(name: String): String? {
        for (i in numberOfParameters - 1 downTo 0) {
            if (name.equals(getParameterName(i), ignoreCase = true)) {
                --numberOfParameters
                names.removeAt(i)
                return values.removeAt(i)
            }
        }
        return null
    }

    fun getParameterName(index: Int): String {
        return names[index]
    }

    fun getParameterValue(index: Int): String {
        return values[index]
    }

    @Throws(JuggleExceptionUser::class)
    fun readParameters(source: String?) {
        val cleanSource = source?.replace("\n", "")?.replace("\r", "") ?: return

        for (token in cleanSource.split(';')) {
            val index = token.indexOf("=")
            if (index > 0) {
                val name = token.take(index).trim()
                val value = token.substring(index + 1).trim()
                if (name.isNotEmpty()) {
                    addParameter(name, value)
                }
            } else {
                val str = token.trim()
                if (str.isNotEmpty()) {
                    val message = jlGetStringResource(Res.string.error_param_has_no_value, str)
                    throw JuggleExceptionUser(message)
                }
            }
        }
    }

    override fun toString(): String {
        val sb = StringBuilder()
        for (i in 0..<numberOfParameters) {
            if (i != 0) {
                sb.append(';')
            }
            sb.append(getParameterName(i)).append('=').append(getParameterValue(i))
        }
        return sb.toString()
    }

    // Throw an appropriate error if there are parameters left over after parsing.

    @Throws(JuggleExceptionUser::class)
    fun errorIfParametersLeft() {
        val count = numberOfParameters
        if (count == 1) {
            val argument = "\"${getParameterName(0)}\""
            val message = jlGetStringResource(Res.string.error_unused_param, argument)
            throw JuggleExceptionUser(message)
        } else if (count > 1) {
            val argument = (0..<count).joinToString(", ") { i -> "\"${getParameterName(i)}\"" }
            val message = jlGetStringResource(Res.string.error_unused_params, argument)
            throw JuggleExceptionUser(message)
        }
    }
}
